### PR TITLE
Fix: DNS NCACHE TTL and OPT RRs

### LINKED ISSUE #2889: [Bug] Too many DNS resolve trials of server domain name.
### Prerequisites

- [X] I understand that this is the official open-source version of the Clash core, **only providing support for the open-source version or Premium version**
- [X] I am submitting an issue with the Clash core, not Clash.Meta / OpenClash / ClashX / Clash For Windows or any other derivative version
- [X] I am using the latest version of the Clash or Clash Premium core **in this repository**
- [X] I have searched at the [Issue Tracker](……/) **and have not found any related issues**
- [X] I have read the [official Wiki](https://dreamacro.github.io/clash/) **and was unable to solve the issue**
- [X] (required for Premium core) I've tried the `dev` branch and the issue still exists

### Version

2023.08.17 Premium

### Operating System

Windows

### Architecture

amd64

### Configuration File

```YAML
proxy-providers:
proxy-groups:
rule-providers:
rules:
  - RULE-SET,BlockedSites,Proxy
  - MATCH,DIRECT
hosts:
dns:
  enable: true
  listen: 127.0.0.1:53
  default-nameserver:
    - 223.6.6.6
    - 119.28.28.28
  enhanced-mode: fake-ip
  fake-ip-range: 198.18.0.1/16
  nameserver:
    - https://dns.alidns.com/dns-query
    - https://doh.pub/dns-query
  fallback:
    - https://1.1.1.1/dns-query
    - https://dns.google/dns-query
  fallback-filter:
    geoip: false
    geoip-code: CN
    ipcidr:
      - 240.0.0.0/4
```

### Log

_No response_

### Description

1. Clash built-in DNS server configured as above. (`geoip: false` is on purpose)

2. SS server with ***ipv6 only*** domain name (quote as `ipv6.server.tld`).

3. As for previous version 0.20.31 (with core ***Premium 2023.07.22***), things went as expected:
- Open CFW, doh server names and ipv6.server.tld resolved and cached, `ipv6.server.tld` got ipv6 address and empty ipv4 address response ([]).
- For later connections, clash won't try to resolve `ipv6.server.tld` to get ipv6 or ipv4 addresses as they have been cached.

4. As for CFW 0.20.32 (with core ***Premium 2023.08.17***), after the initial DNS resolve, clash won't try to resolve `ipv6.server.tld`‘s ipv6 address later. But ***before each*** connection matching THE RULES you know (Global mode is the same), clash will try to resolve `ipv6.server.tld`‘s ipv4 address (A record) and get empty response repeatly .
What as expected is clash don't resolve the A record anymore.

5. Debug logs as below (visit google as example):
- Version ***Premium 2023.07.22*** (No more DNS resolve of ipv6.server.tld):
屏幕截图 2023-08-18 235356

- Version ***Premium 2023.08.17*** (A record resolve trial repeatly; google domains resolved in server side):
屏幕截图 2023-08-18 235304

***RISK***: My server domain get exposed/noticed by ISP as the increase of connections (DNS queried before every connection), as the ISP redirecting DOH server to an unknow site (localhost? ping less than 1ms) sometimes.

I think the core should respect DNS TTL rules (DNS queries that do not yield any results are on purpose usually), or provide an option to do so. I assume, as for the logic of clash usage, valid domain with empty response locally should be added to the RULEs to let server make the query, right?

[Dreamacro/clash#2850](https://github.com/Dreamacro/clash/pull/2850) related, maybe.

### Reproduction Steps

- Clash core 2023.08.17 Premium
- Server with ipv6 only domain name (whatever protocols, ss, vmess etc.)
- DNS server configured
- Make some traffic to the server.

### PR DESCRIPTION
**Pull request by lujjjh**
1. DNS NCACHE was not correctly implemented.
2. OPT RRs must not be cached or forwarded.

Closes #2889.

---

Open question: Should we reintroduce the hack of bypassing caches for ACME challenge RRs?

https://github.com/Dreamacro/clash/blob/c1d027d6d1e6b9db7205ad74a4554c96822110bb/dns/util.go#L37-L41

I'm not sure why we would need this workaround. It appears that the RRs will be cached by the upstream servers regardless.

### CODE REVIEW

**Comment by Dreamacro** [comment]
ACME workaround was added on #2469

**Comment by Dreamacro** [comment]
lint is broken, run `make lint-fix` to fix it

File: dns/util.go [hunk 1]
```diff
@@ -35,7 +38,14 @@ func updateTTL(records []D.RR, ttl uint32) {
 }
 
 func putMsgToCache(c *cache.LruCache, key string, q D.Question, msg *D.Msg) {
-	ttl := minimalTTL(msg.Answer)
+	var ttl uint32
+	if msg.Rcode == D.RcodeServerFailure {
+		// [...] a resolver MAY cache a server failure response.
+		// If it does so it MUST NOT cache it for longer than five (5) minutes [...]
+		ttl = serverFailureCacheTTL
```
[the remark this answers is not in the document]
Comment (lujjjh) [reply dns/util.go:45]: Yeah, I don't want the SERVFAIL to be cached for too long. I just picked 5 seconds randomly.

**Comment by lujjjh** [comment]
> ACME workaround was added on #2469

I'm curious why does this workaround work if the upstream DNS servers also cache records?

**Comment by Dreamacro** [comment]
I think we can keep the original logic in this PR for now.

**Comment by lujjjh** [comment]
> I think we can keep the original logic in this PR for now.

Reintroduced the hack of bypassing caches for ACME challenges.

**Comment by Dreamacro** [comment]
@ffy Would you mind sharing something relevant?

**Comment by ffy** [comment]
> @ffy Would you mind sharing something relevant?

Most ACME tools always perform authoritative resolution, so the main consideration is whether the domain's authoritative NS has the updated record.

cc @lujjjh

**Pull request merged**